indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceInputFormat.java

```diff
@@ -0,0 +1,127 @@
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+import io.druid.data.input.InputRow;
+import io.druid.indexer.HadoopDruidIndexerConfig;
+import io.druid.timeline.DataSegment;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+public class DatasourceInputFormat extends InputFormat<NullWritable, InputRow>
+{
+  private static final Logger logger = new Logger(DatasourceInputFormat.class);
+
+  public static final String CONF_INPUT_SEGMENTS = "druid.segments";
+  public static final String CONF_DRUID_SCHEMA = "druid.datasource.schema";
+  public static final String CONF_MAX_SPLIT_SIZE = "druid.datasource.split.max.size";
+
+  @Override
+  public List<InputSplit> getSplits(JobContext context) throws IOException, InterruptedException
+  {
+    Configuration conf = context.getConfiguration();
+
+    String segmentsStr = Preconditions.checkNotNull(conf.get(CONF_INPUT_SEGMENTS), "No segments found to read");
+    List<DataSegment> segments = HadoopDruidIndexerConfig.jsonMapper.readValue(
+        segmentsStr,
+        new TypeReference<List<DataSegment>>()
+        {
+        }
+    );
+    if (segments == null || segments.size() == 0) {
+      throw new ISE("No segments found to read");
+    }
+
+    logger.info("segments to read [%s]", segmentsStr);
+
+    long maxSize = conf.getLong(CONF_MAX_SPLIT_SIZE, 0);
+
+    if (maxSize > 0) {
+      //combining is to happen, let us sort the segments list by size so that they
+      //are combined appropriately
+      Collections.sort(
+          segments,
+          new Comparator<DataSegment>()
+          {
+            @Override
+            public int compare(DataSegment s1, DataSegment s2)
+            {
+              if (s1.getSize() == s2.getSize()) {
+                return 0;
+              } else if (s1.getSize() < s2.getSize()) {
+                return -1;
+              } else {
+                return 1;
+              }
+            }
+          }
+      );
+    }
+
+    List<InputSplit> splits = Lists.newArrayList();
+
+    List<DataSegment> list = new ArrayList<>();
+    long size = 0;
+
+    for (DataSegment segment : segments) {
+      if (size + segment.getSize() > maxSize && size > 0) {
+        splits.add(new DatasourceInputSplit(list));
+        list = Lists.newArrayList();
+        size = 0;
+      }
+
+      list.add(segment);
+      size += segment.getSize();
+    }
+
+    if (size > 0) {
```

Comment:
if (!list.isEmpty())?

I don't think there will be any size = 0 segments but looking at list directly seems cool